src/core/Cline.ts

```diff
@@ -850,12 +851,22 @@ export class Cline {
 		} catch (error) {
 			// note that this api_req_failed ask is unique in that we only present this option if the api hasn't streamed any content yet (ie it fails on the first chunk due), as it would allow them to hit a retry button. However if the api failed mid-stream, it could be in any arbitrary state where some tools may have executed, so that error is handled differently and requires cancelling the task entirely.
 			if (alwaysApproveResubmit) {
+				const { maxApiRetries } = await this.providerRef.deref()?.getState() ?? {}
+				const currentRetryCount = (this.apiRetryCount || 0) + 1
+				this.apiRetryCount = currentRetryCount
+
+				if (maxApiRetries !== undefined && maxApiRetries !== 0 && currentRetryCount > maxApiRetries) {
+					const errorMsg = `Maximum retry attempts (${maxApiRetries}) reached. ${error.message ?? "Unknown error"}`
+					await this.say("error", errorMsg)
+					throw error
+				}
+
 				const errorMsg = error.message ?? "Unknown error"
 				const requestDelay = requestDelaySeconds || 5
 				// Automatically retry with delay
 				// Show countdown timer in error color
 				for (let i = requestDelay; i > 0; i--) {
-					await this.say("api_req_retry_delayed", `${errorMsg}\n\nRetrying in ${i} seconds...`, undefined, true)
+					await this.say("api_req_retry_delayed", `${errorMsg}\n\nRetrying in ${i} seconds...${maxApiRetries !== undefined && maxApiRetries > 0 ? ` (Attempt ${currentRetryCount} of ${maxApiRetries})` : ''}`, undefined, true)
 					await delay(1000)
 				}
```

Comment:
This is the part I'd be the most careful about when reviewing. If it's wrong/it should be done a different way, I'd love to know!